Add torchvision datasets prepare pipeline

satyaog opened the pull request.

satyaog committed:
Add torchvision datasets prepare pipeline

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -45,56 +58,78 @@ def __call__(
         self,
-        dataset_type: Callable[Concatenate[str, P], VD],
-        relative_paths_to_archives: dict[str, str | Path],
+        source: str | Path = None,
+        files: dict[str, str | Path] = None
```
Use `<type> | None` or `Optional[<type>]` if the default value is `None`, otherwise the type annotation is incorrect.

satyaog replied:
Will do thx!

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -23,15 +24,27 @@
 
 
-class PrepareVisionDataset(Generic[VD, P]):
+class AbstractPrepareVisionDataset(Generic[VD, P]):
```
Why the need for an additional abstract base class? I'm not a big fan of those. Having "abstract" in the name is also a bit ugly IMO.

satyaog replied:
I agree I don't like the name _Abstract_ but it is a known concept I believe so people understand from reading the name what it is. But I will use what you suggested, it's nicer to read

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -146,56 +256,286 @@ def __call__(
+standardized_torchvision_datasets_dirs = {
     Cluster.Mila: "/network/datasets",
-    Cluster.Beluga: "/project/rpp-bengioy/data/curated",
+    Cluster.Beluga: "~/project/rpp-bengioy/data/curated",
```
I'd avoid going through the $HOME filesystem if possible.

satyaog replied:
Indeed we can do what's suggested below

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -45,56 +58,78 @@ def __call__(
-            k: Path(v) for k, v in relative_paths_to_archives.items()
-        }
+        if source is not None:
+            source = Path(source).expanduser()
```
It might be a good idea to resolve symlinks for example `~/scratch` to `/network/scratch/u/username`
```suggestion
            source = Path(source).expanduser().resolve()
```

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -45,56 +58,78 @@ def __call__(
+        if source is not None:
+            source = Path(source).expanduser()
+
+        if source is not None and files is not None:
+            self.relative_paths_to_files = {
+                k: Path(v).relative_to(source) for k, v in files.items()
+            }
```
In what context would it be okay to pass both `source` and `files`?

satyaog replied:
Only when you don't want to have the `relative_to` operation outside of the `SymlinkDatasetFiles`. But creating symlinks is cheap so we could also remove `files` and only use `source`

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -146,56 +256,286 @@ def __call__(
 # only really used to prevent repeating a bit of code in the dictionary below.
 # TODO: Find an exception to this rule and design this dict with that in mind.
-standardized_torchvision_dataset_dirs = {
+standardized_torchvision_datasets_dirs = {
```
nitpick: I disagree on `dataset*s*`:  this means "standardized torchvision dataset directories", hence dataset is singular. 
Actually, perhaps the opposite is better: "standardized_torchvision_datasets_dir"!

satyaog replied:
Ah yes I do prefer `standardized_torchvision_datasets_dir` as well

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -146,56 +256,286 @@ def __call__(
-        )
-        for cluster, dataset_folder in standardized_torchvision_dataset_dirs.items()
+    tvd.Caltech101: {
+        cluster: Compose([
+            SymlinkDatasetFiles(
+                source=f"{datasets_folder}/caltech101",
+            ),
+            MoveFiles(
+                # Torchvision will look into a caltech101 directory to
+                # preprocess the dataset
+                files=[("*", "caltech101/*")]
+            ),
+            PrepareVisionDataset(tvd.Caltech101)
+        ])
```
I really like this!
Perhaps we could rename `PrepareVisionDataset` to just `CallDatasetConstructor` or something like that? That way we could replace the `Abstract(...)` with `PrepareVisionDataset`.

satyaog replied:
Agree

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -45,56 +58,78 @@ def __call__(
+            if entry.is_file():
+                yield entry
+            if entry.is_dir():
+                for _f in SymlinkDatasetFiles._recursive_list_files(entry):
+                    yield _f
```
Nitpick
```suggestion
                yield from SymlinkDatasetFiles._recursive_list_files(entry)
```

satyaog replied:
So much better. I was wondering why it was so ugly to yield a generator in Python. Turns out it was just me

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -104,8 +139,83 @@ def __call__(
+        ----------
+
+        - archives:
+            A list of pair of an archive and a destination's path where the
+            archive should be extracted. The destination's path should be
+            relative.
```
Nitpick:
```suggestion
            A mapping from an archive name to path where the archive
            should be extracted (relative to the 'root' dir).
            The destination paths need to be relative.
```

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -104,8 +139,83 @@ def __call__(
+        ----------
+
+        - files:
+            A list of pair of an archive and a destination's path where the result
+            should be moved and replace. If the destination path's leaf is "*",
+            the destination's parent will be used to hold the file. If not, the
+            destination will be used as the target for the move. The move are
+            executed in sequence. The destination's path should be relative.
```
I'm trying to make this a bit easier to understand, but it's still a bit unclear.
```suggestion
            A mapping from file names to a destination path (relative to a
            'root' dir) where the file should be moved.
            If the destination path ends with "/*" (e.g. "data/*"), the files will
            be moved to the parent folder. Otherwise, the file is moved to
            the destination path.
            Files are moved in sequence. The destination's path need to be relative.
```

satyaog replied:
I'm used to reference a `dict` with _mapping_. You would use _mapping_ as well with a `list` of `tuples`?

lebrice replied:
But in this case, it's a mapping, right? not a list of tuples... Or if it is a list of tuples, then why not use a mapping instead?

lebrice replied:
(since the first item is a string, I don't see why this can't be a dict mapping from `something` -> `something_renamed`)

satyaog replied:
That is true but my brain in that case tends to still reserve some keywords to the type and explain in plain words how the object should be constructed or will be used. Your call :)

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -146,56 +256,286 @@ def __call__(
+                # Torchvision will look for those files to preprocess the
+                # dataset
+                files=[
+                    ("train.tar.gz", "2021_train.tgz"),
+                    ("train_mini.tar.gz", "2021_train_mini.tgz"),
+                    ("val.tar.gz", "2021_valid.tgz"),
```
Just out of curiosity, are those exactly the files torchvision expects? (The checksums for these files will match those expected from torchvision?)

satyaog replied:
Yes and I'm not sure why `torchvision` decided to rename their files actually

lebrice commented on mila_datamodules/cli/prepare_torchvision.py:
```diff
@@ -146,56 +256,286 @@ def __call__(
+    },
+    tvd.Places365: {
+        cluster: Compose([
+            SymlinkDatasetFiles(
+                source=f"{datasets_folder}/places365",
+            ),
+            SymlinkDatasetFiles(
+                source=f"{datasets_folder}/places365.var/places365_challenge",
+            ),
```
These two are happening in the same directory? There's no issue with that?

satyaog replied:
No the files will nicely merge

lebrice reviewed (approved):
Thanks @satyaog !

J'ai mis des suggestions mineures, c'est pas terriblement important.

satyaog committed:
Update mila_datamodules/cli/prepare_torchvision.py

Co-authored-by: Fabrice Normandin <fabrice.normandin@gmail.com>

satyaog committed:
Update mila_datamodules/cli/prepare_torchvision.py

Co-authored-by: Fabrice Normandin <fabrice.normandin@gmail.com>

satyaog committed:
Update mila_datamodules/cli/prepare_torchvision.py

Co-authored-by: Fabrice Normandin <fabrice.normandin@gmail.com>

lebrice committed:
Merge branch 'dev' into feature/preparetv

Signed-off-by: Fabrice Normandin <normandf@mila.quebec>

lebrice committed:
Apply suggestions from code review

Signed-off-by: Fabrice Normandin <normandf@mila.quebec>

satyaog committed:
Small reword

satyaog committed:
Fix prepare torchvision

Fix ExtractArchives, MoveFiles
Stop on dataset load success to avoid redoing extraction and other steps
when dataset is ready

satyaog commented:
Should be good. Tested with Caltech256, CocoCaptions and currently testing with INaturalist

lebrice committed:
Fix bug with cifar10, tiny refactor of a few bits

Signed-off-by: Fabrice Normandin <normandf@mila.quebec>
